nexus/db-queries/src/db/datastore/instance.rs

```diff
@@ -529,4 +530,369 @@ impl DataStore {
+                    );
+                    return Err(Error::internal_error(
+                        "instance is already locked by another saga",
+                    ));
+                }
+                // No saga's ID is set as the instance's `updater_id`. We can
+                // attempt to lock it.
+                None => {}
+            }
+            let gen = instance.runtime_state.updater_gen;
+            slog::debug!(
+                &opctx.log,
+                "attempting to acquire instance updater lock";
+                "instance_id" => %instance.id(),
+                "saga_id" => %saga_lock_id,
+                "current_gen" => ?gen,
+            );
+            (instance, did_lock) = self
+                .instance_updater_try_lock(
+                    opctx,
+                    authz_instance,
+                    gen,
+                    saga_lock_id,
+                )
+                .await?;
+        }
+    }
+
+    /// Attempts to lock an instance's record to apply state updates in an
+    /// instance-update saga, if the pprovided `current_gen` matches the
+    /// instance's current `updater_gen`.
+    ///
+    /// This function will attempt to set the `updater_id` field on the record
+    /// corresponding to the provided `authz_instance`  to the provided
+    /// `saga_lock_id`. If the instance's `updater_gen` field is equal to
+    /// `current_gen`, and the instance's `updater_id` field is null, then the
+    /// generation is advanced and the `updater_id` field is set to
+    /// `saga_lock_id`, acquiring the lock for the calling saga. Otherwise, if
+    /// the generation has advanced since `current_gen` was captured, or if
+    /// another saga has locked the instance, the lock is not acquired.
+    ///
+    /// # Notes
+    ///
+    /// This method MUST only be called from the context of a saga! The
+    /// calling saga must ensure that the reverse action for the action that
+    /// acquires the lock must call [`DataStore::instance_updater_unlock`] to
+    /// ensure that the lock is always released if the saga unwinds.
+    ///
+    /// # Arguments
+    ///
+    /// - `authz_instance`: the instance to attempt to lock to lock
+    /// - `current_gen`: the current generation of the instance's `updater_id`
+    /// - `saga_lock_id`: the UUID of the saga that's attempting to lock this
+    ///   instance.
+    ///
+    /// # Returns
+    ///
+    /// - [`Ok`]`((`[Instance`]`, true))` if the lock was acquired.
+    /// - [`Ok`]`((`[Instance`]`, false))` if the lock was not acquired because
+    ///   the instance was already locked by another saga, or because the
+    ///   generation has advanced since `current_gen` was captured.
+    /// - [`Err`]`(`[`Error::ObjectNotFound`]`)` if no instance with the
+    ///   provided ID exists (or the instance record has been deleted).
+    /// - [`Err`]`(other)` if the database query failed for another reason.
+    pub async fn instance_updater_try_lock(
```

Comment:
I noticed `instance_updater_lock` is pub, and looks like a higher-level interface. Did you want this be pub too?

In particular: This function **does not check** that the value of `updater_id` is currently `None`, so it seems like a caller could invoke this method and overwrite an update lock that is already held.

I think the invocation of this function by `instance_updater_lock` is safe, but that usage could be dangerous if the association of `current_gen` -> `updater_id = None` is not carefully respected

Reply:
Yeah, you're right, this shouldn't be public --- honestly, it can probably just
be folded into the `instance_updater_lock` method. I'll go ahead and do that.